### PR TITLE
chore: Upgrades nextjs to 12.2.X

### LINKED ISSUE #144: Upcoming Next.js middleware changes
In the Next.js package, there is a function which is `withMiddlewareAuth`. In the upcoming breaking changes to Next.js version 12.2, they completely alter how middleware works. I would just like to confirm that this update does not affect this package, as it is a crucial component of my app.

### LINKED ISSUE #166: Sign-in breaks with Next.js 12.2.0
Hi,

Found out when refactoring to use the new Next.js middleware that sign-in just isn't working at all for me in Next.js 12.2.0. Works perfectly when rolling back to next@12.1.6. When signing in, it redirects back to the localhost with the code in the URL, and /api/user is called correctly but /api/auth/callback is never called, thus login never occurs. Have been unable to track down why.

Thanks.

### LINKED ISSUE #168: NextJs - middleware import
### Description

`'@supabase/auth-helpers-nextjs/middleware'` is not found. I used `'@supabase/auth-helpers-nextjs/dist/middleware'` instead. Also, NextJs [replaced the middleware API](https://nextjs.org/docs/messages/nested-middleware) to only have a single root middleware file. [New API Link](https://nextjs.org/docs/advanced-features/middleware). This might have a breaking change for `withMiddlewareAuth`

### PR DESCRIPTION
**Pull request by nickbrinser**
## What kind of change does this PR introduce?

* Updates nextjs to latest (12.2.5).
* Updates usage in middleware example to single `middleware.ts` file, and includes matcher array.
* Updates `access_token` getters and setters for `NextRequest` in `withMiddlwareAuth` and `NextRequestAdapter`.
* Adds export for `./middleware` in nextjs package, allowing removal dist from `withMiddlewareAuth` imports.

These changes aim to fix:
* Fixes #168 
* Fixes #144 
* Fixes #166

### CODE REVIEW

File: examples/nextjs/middleware.ts
Comment (nickbrinser) [inline examples/nextjs/middleware.ts:11]: For now, I opted to stick with the stock matcher using the config object. I like keeping this closer to the NextJS API.

**Review by Aleksion — APPROVED** [review]
Looks good to me. Interesting that NextRequest has changed its cookie interface but NextApiRequest stayed the same.

I'd love to get this in

**Comment by megacherry** [comment]
yeah, me too :)

**Comment by cursecodes** [comment]
It would be great if this could get merged. 12.3 is on the horizon and we still don't have support for 12.2.

**Review by cursecodes — APPROVED** [review]
Looks good to me.

**Comment by kevva** [comment]
Is there anything blocking this?

**Comment by nickbrinser** [comment]
> Is there anything blocking this?

Not that I know of. I just needs a review from a maintainer. @silentworks or @thorwebdev could we get this looked at? I know there are changes in the `next` branch that address this but that relies on the v2 rc as well.

**Comment by thorwebdev** [comment]
Thanks so much @nickbrinser and sorry everyone for the delay, we've got caught up in the RC with the supabase-js V2 changes and dropped the ball on this. 

Let's get this in as `0.2.8` to bridge the gap until we can release the changes on the next branch 👍 

@nickbrinser I've added some changes to your fork here https://github.com/nickbrinser/supabase-auth-helpers/pull/1 . Would you mind merging these in and then I can merge this one here and get it released. Thanks again for your help 💚

**Review by thorwebdev — APPROVED** [review]
LGTM, thanks so much 💚

**Pull request merged**
